### PR TITLE
Test: Unit tests for Member Role

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Issue: #345 

## Changes proposed in this pull request

Added unit test to increase the coverage of MomberRoles.

## Test plan

N/A

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: test/unit/MemberRoles/changeAuthorized.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+describe('changeAuthorized', function () {
+  it('should change authorized address for the role', async function () {
+    const { memberRoles } = this.contracts;
+    const { defaultSender, governanceContracts } = this.accounts;
+
+    const authorizedAddressBefore = await memberRoles.authorized(Role.AdvisoryBoard);
+    await memberRoles.connect(governanceContracts[0]).changeAuthorized(Role.AdvisoryBoard, defaultSender.address);
+    const authorizedAddressAfter = await memberRoles.authorized(Role.AdvisoryBoard);
+
+    expect(authorizedAddressBefore).not.to.be.eq(defaultSender.address);
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeAuthorized.js:13]: minor, but we've been spelling out eq as equal in other tests

File: test/unit/MemberRoles/changeAuthorized.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+const { Role } = require('../utils').constants;
+const { expect } = require('chai');
+
+describe('changeAuthorized', function () {
+  it('should change authorized address for the role', async function () {
+    const { memberRoles } = this.contracts;
+    const { defaultSender, governanceContracts } = this.accounts;
+
+    const authorizedAddressBefore = await memberRoles.authorized(Role.AdvisoryBoard);
+    await memberRoles.connect(governanceContracts[0]).changeAuthorized(Role.AdvisoryBoard, defaultSender.address);
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeAuthorized.js:10]: let's use something else other than `defaultSender.address`

When you write tests, look to make the values arbitrary for generality, whatever that thing that you are switching to is. Arbitrary within constraints of course (eg. arbitrary member)

File: test/unit/MemberRoles/changeAuthorized.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    expect(authorizedAddressBefore).not.to.be.eq(defaultSender.address);
+    expect(authorizedAddressAfter).to.be.eq(defaultSender.address);
+
+    await expect(
+      memberRoles.connect(governanceContracts[0]).changeAuthorized(Role.AdvisoryBoard, governanceContracts[0].address),
+    ).to.be.reverted;
+    await expect(
+      memberRoles.connect(defaultSender).changeAuthorized(Role.AdvisoryBoard, governanceContracts[0].address),
+    ).to.not.reverted;
+  });
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeAuthorized.js:22]: there's no space here

File: test/unit/MemberRoles/changeAuthorized.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    expect(authorizedAddressBefore).not.to.be.eq(defaultSender.address);
+    expect(authorizedAddressAfter).to.be.eq(defaultSender.address);
+
+    await expect(
+      memberRoles.connect(governanceContracts[0]).changeAuthorized(Role.AdvisoryBoard, governanceContracts[0].address),
+    ).to.be.reverted;
+    await expect(
+      memberRoles.connect(defaultSender).changeAuthorized(Role.AdvisoryBoard, governanceContracts[0].address),
+    ).to.not.reverted;
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeAuthorized.js:21]: to.not.be.reverted

File: test/unit/MemberRoles/changeMaxABCount.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    const { memberRoles } = this.contracts;
+    const { governanceContracts } = this.accounts;
+
+    const maxABCountBefore = await memberRoles.maxABCount();
+    await memberRoles.connect(governanceContracts[0]).changeMaxABCount(2);
+    const maxABCountAfter = await memberRoles.maxABCount();
+
+    expect(maxABCountBefore).not.to.be.eq(maxABCountAfter);
+    expect(maxABCountAfter).to.be.eq(2);
+  });
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeMaxABCount.js:14]: please create spaces between test cases

File: test/unit/MemberRoles/changeMaxABCount.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+const { expect } = require('chai');
+
+describe('changeMaxABCount', function () {
+  it('should change max AB count', async function () {
+    const { memberRoles } = this.contracts;
+    const { governanceContracts } = this.accounts;
+
+    const maxABCountBefore = await memberRoles.maxABCount();
+    await memberRoles.connect(governanceContracts[0]).changeMaxABCount(2);
```
Comment (danoctavian) [inline test/unit/MemberRoles/changeMaxABCount.js:9]: use named constants

const newMaxABCount = 2;

File: test/unit/MemberRoles/setup.js [hunk 7]
```diff
@@ -94,13 +94,19 @@ async function setup() {
...
+  // Setting Members
   for (const member of accounts.members) {
     await master.enrollMember(member.address, 1);
     await memberRoles.connect(accounts.governanceContracts[0]).updateRole(member.address, 2, true);
     await nxm.mint(member.address, parseEther('10000'));
     await nxm.connect(member).approve(tokenController.address, parseEther('10000'));
   }
+  // Setting AB Member
+  const [abMember] = accounts.advisoryBoardMembers;
+  await master.enrollMember(abMember.address, 2);
```
Comment (danoctavian) [inline test/unit/MemberRoles/setup.js:106]: use the right constant instead of the 2. Role.whatever

File: test/unit/MemberRoles/stateManagement.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    });
+    expect(memberArray).to.include(abMember.address);
+  });
+
+  it('should return all roles for a member', async function () {
+    const { memberRoles } = this.contracts;
+    const [member] = this.accounts.members;
+
+    const assignedRolesMember = await memberRoles.roles(member.address);
+    expect(assignedRolesMember[0]).to.be.equal(Role.Member);
```
Comment (danoctavian) [inline test/unit/MemberRoles/stateManagement.js:33]: there's no length check here

add length check

File: test/unit/MemberRoles/stateManagement.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  });
+
+  it('should return member at index', async function () {
+    const { memberRoles } = this.contracts;
+    const {
+      members: [member],
+    } = this.accounts;
+
+    const [memberAddress, isActive] = await memberRoles.memberAtIndex(Role.Member, 0);
+    expect(member.address).to.be.equal(memberAddress);
```
Comment (danoctavian) [inline test/unit/MemberRoles/stateManagement.js:67]: this is the other way around.

expect(memberAddress).to.be.equal(member.address)

File: test/unit/MemberRoles/stateManagement.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    expect(isActive).to.be.equal(true);
+  });
+
+  it('should clear storage', async function () {
+    const { memberRoles } = this.contracts;
+
+    await expect(memberRoles.storageCleanup()).to.not.reverted;
```
Comment (danoctavian) [inline test/unit/MemberRoles/stateManagement.js:74]: to.not.be.reverted

File: test/unit/MemberRoles/stateManagement.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      nonMembers: [nonMember],
+      advisoryBoardMembers: [advisoryBoardMember],
+    } = this.accounts;
+
+    const isMemberMember = await memberRoles.checkRole(member.address, Role.Member);
+    const isMemberNonMember = await memberRoles.checkRole(member.address, Role.NonMember);
+    const isMemberABMember = await memberRoles.checkRole(member.address, Role.AdvisoryBoard);
+
+    expect(isMemberMember).to.be.equal(true);
+    expect(isMemberNonMember).to.be.equal(true);
```
Comment (danoctavian) [inline test/unit/MemberRoles/stateManagement.js:90]: this is odd why is this true?

File: test/unit/MemberRoles/stateManagement.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    expect(isNonMemberMember).to.be.equal(false);
+    expect(isNonMemberNonMember).to.be.equal(true);
+    expect(isNonMemberABMember).to.be.equal(false);
+
+    const isABMemberMember = await memberRoles.checkRole(advisoryBoardMember.address, Role.Member);
+    const isABMemberNonMember = await memberRoles.checkRole(advisoryBoardMember.address, Role.NonMember);
+    const isABMemberABMember = await memberRoles.checkRole(advisoryBoardMember.address, Role.AdvisoryBoard);
+
+    expect(isABMemberMember).to.be.equal(true);
+    expect(isABMemberNonMember).to.be.equal(true);
```
Comment (danoctavian) [inline test/unit/MemberRoles/stateManagement.js:106]: also odd that it's true

Comment (danoctavian) [reply test/unit/MemberRoles/stateManagement.js:106]: may be "correct" but warrants at least a comment to clarify why since it looks counterintuitive

**Review by danoctavian — CHANGES_REQUESTED** [review]
some changes necessary

**Review by danoctavian — APPROVED** [review]
LGTM now

File: contracts/mocks/MasterMock.sol [hunk 13]
```diff
@@ -10,8 +10,8 @@ contract MasterMock {
 
   enum Role {
     NonMember,
-    Member,
     AdvisoryBord,
```
Comment (danoctavian) [inline contracts/mocks/MasterMock.sol:13]: mispelling here

**Review by danoctavian — APPROVED** [review]
lgtm but let's fix that mispelling as well since we're here

File: contracts/mocks/MasterMock.sol (hunk 13 above)
Comment (shark0der) [reply contracts/mocks/MasterMock.sol:13]: Let's use `Unassigned` (capitalization) and `AdvisoryBoard` (missing "a").

File: contracts/mocks/MemberRoles/MRMockCover.sol [hunk 14]
```diff
@@ -32,5 +35,7 @@ contract MRMockCover {
   function createMockCover(address to, uint tokenId) public {
     coverNFT.mint(to, tokenId);
   }
-
+  function switchMembershipOf(address member, address newAddress) external {
+    memberRoles.switchMembershipOf(member, newAddress);
+  }
```
Comment (shark0der) [inline contracts/mocks/MemberRoles/MRMockCover.sol:40]: Cover contract doesn't have this function. Why's this here?

File: lib/constants.js [hunk 15]
```diff
@@ -32,7 +32,7 @@ const NXMasterOwnerParamType = {
 };
 
 const Role = {
-  NonMember: 0,
+  UnAssigned: 0,
```
Comment (shark0der) [inline lib/constants.js:35]: `Unassigned`

File: test/unit/Assessment/setup.js [hunk 16]
```diff
@@ -65,8 +66,8 @@ async function setup() {
   const accounts = await getAccounts();
   await master.enrollGovernance(accounts.governanceContracts[0].address);
   for (const member of accounts.members) {
-    await master.enrollMember(member.address, 1); // Uses a different role value than IMemberRoles
-    await memberRoles.enrollMember(member.address, 2); // Uses the actual member role value
+    await master.enrollMember(member.address, Role.Member);
```
Comment (shark0der) [inline test/unit/Assessment/setup.js:69]: This had a value of 1 (AB) previously, why the change?

File: test/unit/MemberRoles/changeDependentContractAddress.js [hunk 17]
```diff
@@ -0,0 +1,45 @@
+const { InternalContractsIDs } = require('../utils').constants;
+const { expect } = require('chai');
+const { hex } = require('../../../lib/helpers');
+const { ZERO_ADDRESS } = require('../../../lib/constants');
```
Comment (shark0der) [inline test/unit/MemberRoles/changeDependentContractAddress.js:4]: Use `AddressZero` from `ethers.constants` instead

File: test/unit/MemberRoles/changeDependentContractAddress.js [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+  before(async function () {
+    const { quotationData, memberRoles } = this.contracts;
+    const { governanceContracts, defaultSender } = this.accounts;
+    await quotationData.connect(governanceContracts[0]).setKycAuthAddress(defaultSender.address);
+    await memberRoles.connect(governanceContracts[0]).setKycAuthAddress(quotationData.address);
+  });
+
+  it('should change authorized address for the role', async function () {
+    const { memberRoles, master } = this.contracts;
+
```
Comment (shark0der) [inline test/unit/MemberRoles/changeDependentContractAddress.js:16]: Given there's only one test here, move the contents of the `before` in the test itself

File: test/unit/MemberRoles/setup.js [hunk 19]
```diff
@@ -94,13 +110,19 @@ async function setup() {
       'Represents all users of Mutual.',
       '0x0000000000000000000000000000000000000000',
     );
-
+  // Setting Members
   for (const member of accounts.members) {
-    await master.enrollMember(member.address, 1);
-    await memberRoles.connect(accounts.governanceContracts[0]).updateRole(member.address, 2, true);
+    await master.enrollMember(member.address, Role.Member);
```
Comment (shark0der) [inline test/unit/MemberRoles/setup.js:115]: was AB

File: test/unit/Assessment/setup.js (hunk 16 above)
Comment (MilGard91) [reply test/unit/Assessment/setup.js:69]: There were different values in the IMember and in MasterMock for the same role as it was in the comments before the fix.

File: lib/constants.js (hunk 15 above)
Comment (MilGard91) [reply lib/constants.js:35]: `UnAssigned` is in the IMemberRoles, that's why it's being used here too

File: contracts/mocks/MemberRoles/MRMockCover.sol (hunk 14 above)
Comment (MilGard91) [reply contracts/mocks/MemberRoles/MRMockCover.sol:40]: `switchMembershipOf` is onlyInternal function in the MemberRoles contract this is being used for testing of that function

File: test/unit/MemberRoles/setup.js (hunk 19 above)
Comment (roxdanila) [reply test/unit/MemberRoles/setup.js:115]: @MilGard91 ?

File: test/unit/MemberRoles/stateManagement.js (hunk 8 above)
Comment (roxdanila) [reply test/unit/MemberRoles/stateManagement.js:34]: @MilGard91

File: test/unit/MemberRoles/stateManagement.js (hunk 11 above)
Comment (roxdanila) [reply test/unit/MemberRoles/stateManagement.js:93]: @MilGard91

File: test/unit/MemberRoles/stateManagement.js (hunk 12 above)
Comment (roxdanila) [reply test/unit/MemberRoles/stateManagement.js:109]: @MilGard91

**Review by roxdanila — CHANGES_REQUESTED** [review]
still open questions

**Review by roxdanila — APPROVED** [review]
